### ISSUE TITLE
lstsq not recovering linear regression parameters using Float32 in Latest Jax/Jaxlib release

### ISSUE THREAD

**Issue by geokwry** [google/jax#13105]
### Description

With the the latest release of Jax/Jaxlib I'm not able to recover parameters in a simple linear model. 

It worked with JAX 0.2.25 and jaxlib 0.1.73
Broke with JAX 0.3.23 and  jaxlib 0.3.22

Using linalg.lstsq

```
import jax.numpy as jnp
n = 10000
X = jnp.transpose(jnp.array([jnp.ones(n), jnp.arange(n)]))
print(X.dtype)
m = 2020.
b = 2020.
weights = jnp.array([b, m], dtype=jnp.float32)
print(weights.dtype)
y = jnp.matmul(X, jnp.transpose(weights))
print(y.dtype)
w = jnp.linalg.lstsq(X, y)[0]
print(w)
# returns [3.0306062e-01 2.0203031e+03]
```

This code also does not cover the original parameters.
```
import jax.numpy as jnp
n = 10000
X = jnp.transpose(jnp.array([jnp.ones(n), jnp.arange(n)]))
print(X.dtype)
m = 2020.
b = 2020.
weights = jnp.array([b, m], dtype=jnp.float32)
print(weights.dtype)
y = jnp.matmul(X, jnp.transpose(weights))
print(y.dtype)
w = jnp.matmul(jnp.matmul(jnp.linalg.inv(jnp.matmul(jnp.transpose(X), X)), jnp.transpose(X)), y)
print(w)
# returns [3717.0405 2019.7451]
```

This seems to be a problem only in float32. When using arrays in float64 I do not observed this issue.

### What jax/jaxlib version are you using?

JAX 0.3.23, jaxlib 0.3.22

### Which accelerator(s) are you using?

CPU

### Additional system info

Linux container running on mac

**Comment by geokwry** [comment]
It's also worth noting that this code returns the original parameters if using numpy instead of jax.numpy.

**Comment by hawkinsp** [comment]
The reason you are getting an inaccurate value is that the smallest singular values of your matrix is smaller than the default `rcond` cutoff. See: https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html

The singular values of `X` are `[5.7730706e+05, 5.0003773e+01]`. The default cutoff used by `jax.numpy.lstsq` is `jnp.finfo(dtype).eps * max(m, n)` and singular values smaller than that value multiplied by the largest singular value are treated as zero for the purposes of rank determination.

The reason you get a different result from NumPy is that NumPY computes `lstsq` in double precision even if you pass it single precision inputs. JAX intentionally does not do this: we never use double precision unless you ask for it.

You will get exactly the same result from classic NumPy if you feed it the same `rcond` value that JAX is using:
`np.linalg.lstsq(X, y, rcond=jnp.finfo(jnp.float32).eps*10000)[0]`
You can try a smaller `rcond` value to recover the smaller singular value, but be careful about losing precision.

For your second example where you are solving the normal equations you have fundamentally the same issue: your matrix is quite badly conditioned for `float32` and you lose accuracy. You should (a) use `float64` by setting `jax.config.update('jax_enable_x64', True)`.

Also, you should not use `inv` to solve a linear system. Try: `jax.scipy.linalg.solve(X.T @ X, X.T @ y, assume_a='pos')`.
